Aquarius/Src/Aquarius/Renderer/VertexArray.h

```diff
@@ -0,0 +1,120 @@
+        Double,
+    };
+
+    static const std::unordered_map<ShaderType, GLenum> ShaderTypeToGLEnum =
+    {
+        {ShaderType::Int, GL_INT},
+        {ShaderType::UInt, GL_UNSIGNED_INT},
+        {ShaderType::Float, GL_FLOAT},
+        {ShaderType::Double, GL_DOUBLE}
+    };
+
+    static const std::unordered_map<ShaderType, uint32_t>  ShaderTypeSizeBytes =
+    {
+            {ShaderType::Int, 4},
+            {ShaderType::UInt, 4},
+            {ShaderType::Float, 4},
+            {ShaderType::Double, 8}
+    };
+
+    class VertexElement
+    {
+    public:
+        VertexElement(ShaderType shaderType, uint32_t count, bool normalize)
+        {
+            m_Type = ShaderTypeToGLEnum.at(shaderType);
+            m_Count = count;
+            m_Size = m_Count * ShaderTypeSizeBytes.at(shaderType);
+            m_Normalize = normalize;
+            m_Offset = 0;
+        }
+
+        void setOffset(uint32_t offset) { m_Offset = offset; }
+
+        uint32_t getOffset() const { return m_Offset; }
+        uint32_t getCount() const { return m_Count; }
+        uint32_t getSize() const { return m_Size; }
+        bool getNormalize() const { return m_Normalize; }
+        GLenum getType() const { return m_Type; }
+
+    private:
+        GLenum m_Type;
+        uint32_t m_Count;
+        uint32_t m_Size;
+        uint32_t m_Offset;
+        bool m_Normalize;
+    };
+
+    class BufferLayout
+    {
+    public:
+        BufferLayout() = delete;
+        ~BufferLayout() = default;
+        BufferLayout(std::initializer_list<VertexElement> vertexElements)
+        {
+            // Calculate offset for each element
+            uint32_t Offset = 0;
+            for (auto elem : vertexElements)
+            {
+                elem.setOffset(Offset);
+                m_Elements.push_back(elem);
+                Offset += elem.getSize();
+            }
+
+            m_Stride = Offset;
+        }
+
+        const std::vector<VertexElement>& getElements() const { return m_Elements; }
+        uint32_t getStride() const { return m_Stride; }
+    private:
+        std::vector<VertexElement> m_Elements;
+        uint32_t m_Stride;
+    };
+
+    class VertexArray
+    {
+    public:
+        VertexArray();
+        VertexArray(const std::shared_ptr<VertexBuffer>& vertexBuffer,
+                    const std::shared_ptr<IndexBuffer>& IndexBuffer,
+                    const BufferLayout& bufferLayout);
+        ~VertexArray();
+
+        void activate();
+        void deactivate();
+
+        void setVertexBuffer(const std::shared_ptr<VertexBuffer>& vertexBuffer);
+        void setVertexBuffer(const std::shared_ptr<VertexBuffer>& vertexBuffer, const BufferLayout& bufferLayout);
+        void setIndexBuffer(const std::shared_ptr<IndexBuffer>& IndexBuffer);
+        void setBufferLayout(const BufferLayout& bufferLayout);
+
+        std::shared_ptr<VertexBuffer> getVertexBuffer() const;
+        std::shared_ptr<IndexBuffer> getIndexBuffer() const;
```

Comment:
In the same vein, do we want to be definitely incrementing the refcount here when someone wants to view/modify the vertex/index buffers? We may wanna just return a raw pointer here

Reply:
If this were multi threaded, it would be dangerous to use a raw pointer here I think because the ref count could hit 0 while something else is using the raw pointer obtained from these methods. In our case, everything occurs sequentially so I think we may be good with raw? Unless the renderer for some reason has to wait until the next frame to render or something. Its hard to know what to do until the implementation details come around

Reply:
Yeah that's a fair point, well in that case i'd say its ok to leave the way it is, but your call. Idt we will get around to threading anyways

Reply:
Yeah we definitely wont, just wondering what the negative to increasing the ref count is? I guess if something just wants a peak at the VB/IB it doesnt need/want to be part owner

Reply:
Yeah exactly, so there's no real harm other than a slight performance hit (obviously the refcount has to be checked then when that returned shared_ptr goes out of scope to determine whether or not to deallocate the memory, and it may only be in scope for a very short period of time)

Reply:
raw pointer it is